pinot-plugins/pinot-file-system/pinot-adls/src/main/java/org/apache/pinot/plugin/filesystem/ADLSGen2PinotFS.java

```diff
@@ -106,24 +118,75 @@ public void init(PinotConfiguration config) {
     String blobServiceEndpointUrl = HTTPS_URL_PREFIX + accountName + AZURE_BLOB_DNS_SUFFIX;
 
-    StorageSharedKeyCredential sharedKeyCredential = new StorageSharedKeyCredential(accountName, accessKey);
+    DataLakeServiceClientBuilder dataLakeServiceClientBuilder = new DataLakeServiceClientBuilder().endpoint(dfsServiceEndpointUrl);
+    BlobServiceClientBuilder blobServiceClientBuilder = new BlobServiceClientBuilder().endpoint(blobServiceEndpointUrl);
+
+    if (accountName!= null && accessKey != null) {
+      LOGGER.info("Authenticating using the access key to the account.");
+      StorageSharedKeyCredential sharedKeyCredential = new StorageSharedKeyCredential(accountName, accessKey);
+      dataLakeServiceClientBuilder.credential(sharedKeyCredential);
+      blobServiceClientBuilder.credential(sharedKeyCredential);
+    } else if (clientId != null && clientSecret != null && tenantId != null) {
+      LOGGER.info("Authenticating using Azure Active Directory");
+      ClientSecretCredential clientSecretCredential = new ClientSecretCredentialBuilder()
+          .clientId(clientId)
+          .clientSecret(clientSecret)
+          .tenantId(tenantId)
+          .build();
+      dataLakeServiceClientBuilder.credential(clientSecretCredential);
+      blobServiceClientBuilder.credential(clientSecretCredential);
+    } else {
+      // Error out as at least one mode of auth info needed
+      throw new IllegalArgumentException("Expecting either (accountName, accessKey) or (clientId, clientSecret, tenantId)");
+    }
 
-    DataLakeServiceClient serviceClient = new DataLakeServiceClientBuilder().credential(sharedKeyCredential)
-        .endpoint(dfsServiceEndpointUrl)
-        .buildClient();
+    _blobServiceClient = blobServiceClientBuilder.buildClient();
+    DataLakeServiceClient serviceClient = dataLakeServiceClientBuilder.buildClient();
+    _fileSystemClient = getOrCreateClientWithFileSystem(serviceClient, fileSystemName);
 
-    _blobServiceClient =
-        new BlobServiceClientBuilder().credential(sharedKeyCredential).endpoint(blobServiceEndpointUrl).buildClient();
-    _fileSystemClient = serviceClient.getFileSystemClient(fileSystemName);
     LOGGER.info("ADLSGen2PinotFS is initialized (accountName={}, fileSystemName={}, dfsServiceEndpointUrl={}, "
             + "blobServiceEndpointUrl={}, enableChecksum={})", accountName, fileSystemName, dfsServiceEndpointUrl,
         blobServiceEndpointUrl, _enableChecksum);
   }
 
+  /**
+   * Returns the DataLakeFileSystemClient to the specified file system creating if it doesn't exist.
+   *
+   * @param serviceClient authenticated data lake service client to an account
+   * @param fileSystemName name of the file system (blob container)
+   * @return DataLakeFileSystemClient with the specified fileSystemName.
+   */
+  @VisibleForTesting
+  public DataLakeFileSystemClient getOrCreateClientWithFileSystem(DataLakeServiceClient serviceClient,
```

Comment:
I think that we should throw the exception instead of creating the file system in Azure. We can assume that the blob container is created beforehand because this file system may need to be configured differently. Auto-managing containers will be handy if we need to use multiple containers but in our case, we just need a single container with ADLS Gen2 feature enabled.

Please double check Google Cloud Storage and AWS S3 implementation. We should keep them in sync.

Reply:
Its consistent with the gen1 implementation as the blob is an inclusion in gen2. the file paths get created if there are none, and this is following the same. Auto create makes the blob concept transparent to user (like gen1)

Reply:
@rkanumul  I'm not sure if this is consistent with gen1 implementation. Your implementation would create the new file system resource for the Azure account and this is not what's happening in Gen1 implementation. Am I missing something?

I am against of Pinot code creating the resources on the Azure account. ADLS Gen2 needs to be pre-created and hooked with the Pinot.

Reply:
gen1 doesn't have the concept of "filesystem" which is nothing but blob container.
So in gen1 one would just create a top level directory.
That is an implementation detail for this resource and is not to be treated as separate from filepath itself.
**It is similar to create a directory path if it doesn't exist in gen1.** ans should not be treated similar to account creation but file creation.

The reason they implemented it in this way is they had blob store with all the file system semantics already. they just converged them.
https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction

https://docs.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-namespace

Reply:
@rkanumul If Gen1 code automatically creates the top-level blob container, I think that we should change that. I still think that our code should not create the top-level container and we should assume that they are pre-generated. Pinot should be able to generate any files/directories within the container.

Reply:
There is some misunderstanding here, gen1 doesn't have the concept of blob store. I was equating gen2 blob with creating the directories in gen1. eg how we create pinot_segment/<cluster-name> automatically even though they are not pre created.
I see what you are saying but don't quite agree with that. Sure, they can create/update it if they have particular need, but we provide a default.
Don't agree with the "clean unintentionally created" part, that can happen to even directories with clustername or some other auto created path.

Reply:
@rkanumul I went over the Azure portal. I actually got confused between `storage account` and `container`.  `storage account` looks to be the top-level storage resource concept in Azure while `bucket` is the top level concept in AWS s3. I thought that `container` in Azure is an equivalent concept to `bucket` in AWS. 

So yes, I actually agree with you that we can create the container here in the code. It's equivalent to creating a directory within the root path.

Sorry for the misunderstanding.

Reply:
yes, that is what I thought.
No worries, I appreciate the detail in review. :)